File: simulate.jl
#!/home/david/julia-1.6.1/bin/julia
#import GZip
if !@isdefined(CommandLine_loaded)
  include("CommandLine.jl")
end
# if !@isdefined(sort_loaded)
#   include("sort.jl")
# end
if !@isdefined(util_loaded)
  include("util.jl")
end
if !@isdefined(sort_loaded)
  include("sort.jl")
end

using Printf
using DelimitedFiles
using Random

mutable struct HawkesPoint
  process::Int64
  time::Float64
end

function Base.:<(x::HawkesPoint,y::HawkesPoint)
  return x.time < y.time
end

mutable struct Process
  process_no::Int64
  parent::Int64
  generation::Int64
  start_time::Float64
end

function main(cmd_line = ARGS)    
  defaults = Dict{String,Any}(
    "seed" => 12345,
    "file_dir" => "",
    "out_file"=>"hawkes_test_data.txt",
    "total_time" => 10,
    "lambda_0" => 1, # base process rate
    "sigma_0" => 2, # initial child process rate (see below)
    "child_half_life" => .1, #half-life of child processes 
  )
  cl = get_vals(defaults,cmd_line) # replace defaults with command line values if they are specified
  println("parameters: $defaults")
  seed = defaults["seed"]
  out_file = defaults["out_file"]
  total_time = defaults["total_time"]
  lambda_0 = defaults["lambda_0"]
  sigma_0 = defaults["sigma_0"]
  child_half_life = defaults["child_half_life"]

  data = HawkesPoint[]
  processes = Process[]
  next_process::Int64 = 0 #indexes the processes array
  next_index::Int64 = 0 #indexes the data array
  rng = MersenneTwister(seed)
  t = 0
  i = 0
  
  parent_process = Process(0,-1,-1,0) # dummy parent data for base process
  data_point = HawkesPoint(0,0)
  rho = 0 #fake values for base process
  sigma = lambda_0
  println("dummy parent for base process: $parent_process")
  
  while next_index <= length(data)
    if(next_index > 0) # skip the startup for the base process
      data_point = data[next_index]
      parent_process = processes[data_point.process]
      rho = log(2)/child_half_life # decay rate of child lambda
      sigma = sigma_0
    end
     # OK, we have a new parent. Set up the child process (this will be the base on the first time thru
    child_process = Process(length(processes)+1,parent_process.process_no,parent_process.generation+1,data_point.time)
    push!(processes,child_process) # save the child process
    t0 = child_process.start_time
    t = t0
    while true # generate the next point
      lambda = sigma*exp(-rho*(t-t0))/(2^(child_process.generation))
      y = rand(rng)
      t -= log(1-y)/lambda;
      if t > total_time;break;end
      push!(data,HawkesPoint(child_process.process_no,t))
    end # while true
    next_index += 1
  end # while(next_index <= length(data)
  heapsort(data)
  #  println(out_file,"data:\n")
  open(out_file, "w") do io
    for p in data
      write(io, "$(p.process)  $(p.time)\n")
    end
  end
  

  println("processes:\n")
  for p in processes
    println("$(p.process_no): parent: $(p.parent) generation: $(p.generation)  start_time: $(p.start_time)")
    end
end #main


# execution begins here
if occursin("simulate.jl",PROGRAM_FILE)
  #was simulate.jl called by a command?
  println("calling main with ARGS = $ARGS")
  main(ARGS) 
else
  if isinteractive()
    print("enter command line: ")
    cmd = readline()
    main(map(string,split(cmd)))
  end
end 


  #  while true
  #     t -= log(1-rand(rng))/lambda_0;
  #     i += 1
  #     if t > total_time;break;end
  #     push!(data,HawkesPoint(next_process,parent,false,t))
  #     println("$i: $t")
  #   end
  #   next_process += 1
  # while i > 1
  #   i -= 1
  #   println(data[i])
  # end
 

File: simulate.cc
#include<iostream>
#include<fenv.h>
#include<fstream>
#include<cmath>
#include<string>
#include "using.h"
#include "Awk.h"
#include "Array.h"
#include "Matrix.h"
#include "GetOpt.h"
#include "util.h"
#include "stats.h"
#include "FlexHeap.h"

struct Process{
  int process_no; // index in the processes table
  int parent_no; // index of the parent in the processes table
  int generation; // parent genration + 1
  int seq_no; // sequence no. in the current child process
  double parent_start;
  Process(int n=0, int p=-1, int g=0, int s = 0, double t=0.0):
    process_no(n),parent_no(p),generation(g),seq_no(s), parent_start(t) {}
  Process& operator=(const Process& p){
    process_no = p.process_no;
    parent_no = p.parent_no;
    generation = p.generation;
    seq_no = p.seq_no;
    parent_start = p.parent_start;
    return *this;
  }
};

ostream& operator<<(ostream& os, const Process& p){
  os << format("process no. %d: parent_no: %d generation: %d sequence_no: %d start: %f",
               p.process_no,p.parent_no,p.generation,p.seq_no,p.parent_start);
  return os;
}

struct HawkesPoint{
  Process mark; // mark is the generating process for this event
  double time;
  bool operator< (const HawkesPoint& h) const { return time < h.time;}
  HawkesPoint(Process m = 0, double t = 0.0) : mark(m), time(t) {}
  HawkesPoint& operator=(const HawkesPoint& h){
    mark = h.mark;
    time = h.time;
    return *this;
  }
};

ostream& operator<<(ostream& os, const HawkesPoint& h){
  os << format("mark: %s time: %f\n",h.mark,h.time);
  return os;
}
  
 
int main(int argc, char** argv){

  string model_input_file = "";
  string output_file = "hawkes_test_data.txt";
  string file_dir = "";
  int ndata = 100;
  double sigma_0 = 2;
  double half_life = 2;
  double lambda_0 = 1;
  double max_time = 5;
  int seed = 12345;
  
  GetOpt cl(argc,argv); // parse command line
  cl.get("output_file",output_file); cout << "model output to "<<output_file<<endl;
  cl.get("file_dir",file_dir); cout << "file directory is "<<file_dir<<endl;
  cl.get("ndata",ndata); cout << "ndata: "<<ndata<<endl;
  cl.get("sigma_0",sigma_0); cout << "sigma_0: "<<sigma_0<<endl;
  cl.get("half_life",half_life); cout << "half_life: "<<half_life<<endl;
  cl.get("lambda_0",lambda_0); cout << "lambda_0: "<<lambda_0<<endl;
  cl.get("max_time",max_time); cout << "max_time: "<<max_time<<endl;
  cl.get("seed",seed); cout << "seed: "<<seed<<endl;
  
  double rho = log(2)/half_life; // decay rate of child processes
  feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
  if(file_dir !=""){
    model_input_file = file_dir+"/"+model_input_file;
    output_file = file_dir+"/"+output_file;
  }
  std::ofstream output(output_file.c_str());
  if(output_file != "" && !output.good()){
    cerr << "can't open "<< output_file<<endl;
    exit(1);
  }
  
  FlexHeap<HawkesPoint> data(10000);
  //  Array<Process> processes(10000);
  int nparents = 0; // no. of parents processed so far
  int npoints = 0; // no. of data points stored so far (store next data point at data[npoints]
  Process fake; // fake parent for base process
  data[npoints++] = HawkesPoint();
  //  processes[0] = fake;
  LFSR128 rng(seed);
  
  while(nparents < npoints){
    Process& parent = data[nparents].mark; // the child becomes a parent
    double t0 = data[nparents].time;
    Process& this_process = data[npoints].mark;
    if(nparents > 0) this_process = Process(nparents,parent.process_no,parent.generation+1,0,t0);
    else this_process = fake;
    cout << "new process: "<<this_process <<" time: "<<t0<<endl;
    //    processes[nparents] = this_process;
    double t = t0;
    while(true){
      double sigma = sigma_0;
      double lambda = this_process.process_no == 0? lambda_0 : sigma*exp(-rho*(t-t0))/pow(2,this_process.generation);
      t -= log(1-rng.uniform())/lambda; // delta_t is exponentially distributed with rate lambda
      if(t > max_time) break;
      this_process.seq_no++;
      data[npoints++] = HawkesPoint(this_process,t);
    }
    cout << format("process %d has %d points\n",this_process.process_no,this_process.seq_no);
    nparents++;
  } // OK, create the next child process
  data.sort(npoints);
  cout << "sorted (but not re-indexed) data:\n";
  for(int i = 0;i < npoints;i++) cout << data[i].mark<<" "<<data[i].time<<endl;
  Array<int> new_process_no(10000);
  new_process_no.fill(-1);
  int new_i;
  for(int i = 0;i < npoints-1;i++){ // re-number processes after sorting
    Process m = data[i+1].mark;
    if(new_process_no[m.process_no] == -1) new_process_no[m.process_no] = i; // start of new process
    int pno = (m.parent_no >= 0 ? m.parent_no : 0);
    output<<new_process_no[pno]<<" "<<data[i+1].time<<endl;
  }
  output.close();
  for(int i = 0;i < new_process_no.len();i++){
    cout << format("%d -> %d\n",i,new_process_no[i]);
  }
}  
